From 9bce174e309be04770886b0009cdbaffb1ab1a2a Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:52:19 -0400
Subject: [tools] perf/diff: Color the Weighted Diff column

Message-id: <1408953210-26343-252-git-send-email-jolsa@redhat.com>
Patchwork-id: 89001
O-Subject: [PATCH RHEL7.1 BZ1133083 251/322] perf diff: Color the Weighted Diff column
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit a5846e215bd47f61133383822422c683600efa7a
Author: Ramkumar Ramachandra <artagnon@gmail.com>
Date: Mon Dec 30 13:32:35 2013 +0530

description
===========
In

  $ perf diff -c wdiff:M,N

color the numbers in the Weighted Diff column using color_snprintf(),
picking the colors using get_percent_color().
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-diff.c b/tools/perf/builtin-diff.c
index 73d8bff..a77e312 100644
--- a/tools/perf/builtin-diff.c
+++ b/tools/perf/builtin-diff.c
@@ -778,6 +778,7 @@ static int __hpp__color_compare(struct perf_hpp_fmt *fmt,
   container_of(fmt, struct diff_hpp_fmt, fmt);
  struct hist_entry *pair = get_pair_fmt(he, dfmt);
  double diff;
+ s64 wdiff;
  char pfmt[20] = " ";
 
  if (!pair)
@@ -806,6 +807,18 @@ static int __hpp__color_compare(struct perf_hpp_fmt *fmt,
   scnprintf(pfmt, 20, "%%%d.6f", dfmt->header_width);
   return value_color_snprintf(hpp->buf, hpp->size,
      pfmt, diff);
+ case COMPUTE_WEIGHTED_DIFF:
+  if (he->dummy)
+   goto dummy_print;
+  if (pair->diff.computed)
+   wdiff = pair->diff.wdiff;
+  else
+   wdiff = compute_wdiff(he, pair);
+
+  scnprintf(pfmt, 20, "%%14ld", dfmt->header_width);
+  return color_snprintf(hpp->buf, hpp->size,
+    get_percent_color(wdiff),
+    pfmt, wdiff);
  default:
   BUG_ON(1);
  }
@@ -826,6 +839,12 @@ static int hpp__color_ratio(struct perf_hpp_fmt *fmt,
  return __hpp__color_compare(fmt, hpp, he, COMPUTE_RATIO);
 }
 
+static int hpp__color_wdiff(struct perf_hpp_fmt *fmt,
+   struct perf_hpp *hpp, struct hist_entry *he)
+{
+ return __hpp__color_compare(fmt, hpp, he, COMPUTE_WEIGHTED_DIFF);
+}
+
 static void
 hpp__entry_unpair(struct hist_entry *he, int idx, char *buf, size_t size)
 {
@@ -1007,6 +1026,9 @@ static void data__hpp_register(struct data__file *d, int idx)
  case PERF_HPP_DIFF__RATIO:
   fmt->color = hpp__color_ratio;
   break;
+ case PERF_HPP_DIFF__WEIGHTED_DIFF:
+  fmt->color = hpp__color_wdiff;
+  break;
  default:
   break;
  }
-- 
1.7.1